crates/red_knot_test/src/parser.rs

```diff
@@ -0,0 +1,467 @@
+
+static CODE_RE: Lazy<Regex> = Lazy::new(|| {
+    Regex::new(r"^```(?<lang>\w+)(?<config>( +\S+)*)\s*\n(?<code>(.|\n)*?)\n```\s*\n").unwrap()
+});
+
+static IGNORE_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"^.*\n?").unwrap());
+
+#[derive(Debug)]
+struct Parser<'s> {
+    /// Sections of the final [`MarkdownTestSuite`].
+    sections: IndexVec<SectionId, Section<'s>>,
+
+    /// Files of the final [`MarkdownTestSuite`].
+    files: IndexVec<FileId, File<'s>>,
+
+    /// The unparsed remainder of the Markdown source.
+    rest: &'s str,
+
+    /// Stack of ancestor sections.
+    stack: Vec<SectionId>,
+
+    /// True if current active section (last in stack) has files.
+    current_section_files: Option<FxHashSet<&'s str>>,
+}
+
+impl<'s> Parser<'s> {
+    fn new(path: &'s str, source: &'s str) -> Self {
+        let mut sections = IndexVec::default();
+        let root_section_id = sections.push(Section {
+            title: path,
+            level: 0,
+            parent_id: None,
+        });
+        Self {
+            sections,
+            files: IndexVec::default(),
+            rest: source,
+            stack: vec![root_section_id],
+            current_section_files: None,
+        }
+    }
+
+    fn parse(mut self) -> anyhow::Result<MarkdownTestSuite<'s>> {
+        self.start()?;
+        Ok(self.finish())
+    }
+
+    fn finish(mut self) -> MarkdownTestSuite<'s> {
+        self.sections.shrink_to_fit();
+        self.files.shrink_to_fit();
+
+        MarkdownTestSuite {
+            sections: self.sections,
+            files: self.files,
+        }
+    }
+
+    fn start(&mut self) -> anyhow::Result<()> {
+        while !self.rest.is_empty() {
+            if let Some(caps) = self.scan(&HEADER_RE) {
+                let header_level = caps["level"].len();
+                self.pop_sections_to_level(header_level);
+
+                // We never pop the implicit root section.
+                let parent = self.stack.last().unwrap();
+
+                let section = Section {
+                    // HEADER_RE can't match without a match for group 'title'.
+                    title: caps.name("title").unwrap().into(),
+                    level: header_level,
+                    parent_id: Some(*parent),
+                };
+
+                if self.current_section_files.is_some() {
+                    return Err(anyhow::anyhow!(
+                        "Header '{}' not valid inside a test case; parent '{}' has code files.",
+                        section.title,
+                        self.sections[*parent].title,
+                    ));
+                }
+
+                let section_id = self.sections.push(section);
+                self.stack.push(section_id);
+
+                self.current_section_files = None;
+            } else if let Some(caps) = self.scan(&CODE_RE) {
+                // We never pop the implicit root section.
+                let parent = self.stack.last().unwrap();
```

Comment:
the fact that you have to do this `.last().unwrap()` thing three times in this file, and have to have the same comment next to two of those times explaining why it's safe, makes me wonder if you should create a newtype wrapper for `self.stack`:

```rs
struct SectionStack(Vec<SectionId>);

impl SectionStack {
    fn parent(&self) -> &SectionId {
        self.0.last().expect("Should never pop the implicit root section")
    }
```